src/com/amazon/ionschema/internal/SchemaImpl.kt

```diff
@@ -129,35 +128,51 @@ internal class SchemaImpl private constructor(
                            header: IonStruct): Map<String, Import> {
 
         val importsMap = mutableMapOf<String, SchemaAndTypeImports>()
+        val importSet: MutableSet<String> = schemaSystem.getSchemaImportSet()
+
         (header.get("imports") as? IonList)
             ?.filterIsInstance<IonStruct>()
             ?.forEach {
                 val id = it["id"] as IonText
-                val importedSchema = schemaSystem.loadSchema(id.stringValue())
-                val schemaAndTypes = importsMap.getOrPut(id.stringValue()) {
-                    SchemaAndTypeImports(id.stringValue(), importedSchema)
-                }
+                val alias = it["as"] as? IonSymbol
+                // if importSet has an import with this id then do not load schema again to break the cycle.
+                if(!importSet.contains(id.stringValue())) {
+                    var normalizedId = schemaId?: "";
+
+                    // if Schema is importing itself then throw error
+                    if(normalizedId.equals(id.stringValue())) {
+                        throw InvalidSchemaException("Schema can not import itself.")
+                    }
 
-                val typeName = (it["type"] as? IonSymbol)?.stringValue()
-                if (typeName != null) {
-                    var newType = importedSchema.getType(typeName)
-                            ?: throw InvalidSchemaException(
-                                "Schema $id doesn't contain a type named '$typeName'")
+                    // add parent and current schema to importSet and continue loading current schema
+                    schemaSystem.addToSchemaImportSet(normalizedId)
+                    schemaSystem.addToSchemaImportSet(id.stringValue())
+                    val importedSchema = schemaSystem.loadSchema(id.stringValue())
+                    schemaSystem.resetSchemaImportSet(id.stringValue(), normalizedId)
```

Comment:
There's a small incongruity here: we add the parent ID and child ID as two separate calls:

```kotlin
schemaSystem.addToSchemaImportSet(normalizedId)
schemaSystem.addToSchemaImportSet(id.stringValue())
```

but then remove them in a single call:

```kotlin
schemaSystem.resetSchemaImportSet(id.stringValue(), normalizedId)
```

Do we need the helper methods at all? Since it's an internal-only data structure, could we just use: 
```kotlin
importSet.add(parentId);
importSet.add(childId);
// ... do things
importSet.remove(parentId);
importSet.remove(childId);
```
?

Reply:
Do you mean setting importset to be `public` and directly using its `add` and `remove` methods?

Reply:
I don't think you need to make it `public`, do you? You [already have a handle on it](https://github.com/amzn/ion-schema-kotlin/blob/a5119e30891c291a09b4645c00b00c21bbb33b80/src/com/amazon/ionschema/internal/SchemaImpl.kt#L131):

```kotlin
val importSet: MutableSet<String> = schemaSystem.getSchemaImportSet()
```